Address Sphinx 2 compatibility issues

agjohnson opened the pull request:
The main change here is that we load information into the template so
that CSS can make decisions about what element hierarchy it can expect,
instead of covering both html4 and html5 writer with a single set of
rules. I also fixed the tox config so that it generates docs for
testing.

This is also in part going to revert the changes in #785. Instead of
maintaining a list of `dl` classes to apply the domain styles to, this
PR instead maintains a list of definition list classes to *not* apply
the domain definition list styling too. There are too many possibilities
to cover with an include list.

Fixes #800
Fixes #801 
Fixes #803
Fixes #831
Fixes #802  
Fixes #798 
Fixes #799 
Fixes #741 
Fixes #746 
Fixes #855 

### Note for issues that will be closed by this PR

I will be cutting a development release for the theme as soon as this PR is merged. We could use some testing at that point for the release `0.5rc1`, which will be available on PyPI (hopefully) early next week.

jessetan commented on sphinx_rtd_theme/layout.html:
```diff
@@ -7,10 +7,11 @@
   {%- set titlesuffix = "" %}
 {%- endif %}
 {%- set lang_attr = 'en' if language == None else (language | replace('_', '-')) %}
+{%- set sphinx_writer = 'writer-html5' if html5_doctype else 'writer-html4' %}
```
An alternative would be to use a `sphinx2` class. If we ever have similar styling incompatibilities with e.g. Sphinx 2.6 or 3.0, they will still use the HTML5 writer and we will need a different class to differentiate.

agjohnson replied:
True, I could see the writer output changing in the future and we might need more granularity eventually. In this case, you _can_ still use the html 4 writer with Sphinx 2, so the most truthy check would be the writer class (`html5_doctype` seemed to be the closest option already in the context data passed by sphinx).

When we cross it, I could see us needing logic like that also uses the sphinx version (assuming there is some major backwards incompatible change at some point)

agjohnson commented:
Added a method to also test the badge only css and I found a maybe sphinx 2 related bug. I added a switcher debug link:

![image](https://user-images.githubusercontent.com/1140183/68353874-11c7fe00-00c8-11ea-9c74-22909bb49890.png)

jessetan commented on tox.ini:
```diff
@@ -15,18 +15,17 @@ deps =
-commands =
-    sphinx-build -b html -d {envtmpdir}/doctrees .  {envtmpdir}/html
+    py.test {posargs} tests/
+    sphinx-build -b html -d {envtmpdir}/doctrees docs/ {envtmpdir}/html
```
Building the docs with all these versions of Sphinx fails the tests. It looks like this theme has not been able to build its own documentation with Sphinx < 1.5 for some time.

I can get some tests working by changing the first line of `setup()` in `sphinx_rtd_theme/__init__.py` to:
```py
if sphinx.version_info >= (1, 6, 0):
        app.add_html_theme('sphinx_rtd_theme', path.abspath(path.dirname(__file__)))
```

After that change, the sphinx13 and sphinx14 tests fail with
```
sphinx_rtd_theme/.tox/py27-sphinx13/lib/python2.7/site-packages/sphinx/locale/__init__.py", line 195, in _
    return translators['sphinx'].ugettext(message)
KeyError: 'sphinx'
```
 I suggest to revert these lines and separately address compatibility with Sphinx 1.3 and 1.4 (or remove them from the test), so this PR can land.

agjohnson replied:
We've discussed removing support for old versions of Sphinx, there likely is not a lot of benefit to supporting these old versions. Even 1.5 is 4 years old at this point.

Blendify commented:
@agjohnson can you please update this patch so the tests pass?

davidelbaze commented:
up ? ^^

agjohnson commented:
I am currently finishing up this PR and I will be breaking out some of the unrelated work into separate PRs.

I am resolving the remaining issue with field lists, definition lists, option lists, and footnotes with CSS `grid`. This is a fairly well supported option and it works more consistently than flex box.

agjohnson commented:
Rebased and should be ready for final review! I've enabled PR builder, though the assets look cached at the moment. PR builds should use the locally available theme for testing, though we should probably be using `?{{ commit }}` or something similar to avoid caching the static assets through each build.

davidfischer commented on src/sass/_theme_rst.sass:
```diff
@@ -239,22 +242,63 @@
+    dl.footnote,
+    dl.field-list
+      display: grid
+      grid-template-columns: max-content auto
```
```suggestion
      grid-template-columns: max-content auto
      > dt
        padding-left: 1rem
```

There is a small amount of left padding on the field lists in the HTML4 writer table output.

agjohnson replied:
Oh yeah. I left this one because i think it looks better without the left padding. Any thoughts on which to use?

agjohnson replied:
We can come back to this later, I'll just match the output.

agjohnson commented:
Thanks for review everyone! I'll move on to cutting a release candidate this afternoon.
